From 03178b4ed8f42cc5e2d4c49440c72ebf937eff1d Mon Sep 17 00:00:00 2001
From: Andrew Eikum <aeikum@codeweavers.com>
Date: Wed, 26 Feb 2020 11:08:35 -0600
Subject: [PATCH] steam_helper: Don't use msvcrt

We need access to linux crt functions (like getenv()), so disallow use
of msvcrt. Notably, linux crt wchar functions are _not_ compatible with
Windows WCHAR strings.
---
diff --git a/steam.cpp b/steam.cpp
index 8e5b977..37e7568 100644
--- a/steam.cpp
+++ b/steam.cpp
@@ -117,9 +117,19 @@ static void setup_steam_registry(void)
     SteamAPI_Shutdown();
 }
 
+static WCHAR *strchrW(WCHAR *h, WCHAR n)
+{
+    while(*h)
+    {
+        if(*h == n) return h;
+        ++h;
+    }
+    return NULL;
+}
+
 static WCHAR *find_quote(WCHAR *str)
 {
-    WCHAR *end = wcschr(str, '"'), *ch;
+    WCHAR *end = strchrW(str, '"'), *ch;
     int odd;
     while (end)
     {
@@ -132,7 +142,7 @@ static WCHAR *find_quote(WCHAR *str)
         }
         if (!odd)
             return end;
-        end = wcschr(end + 1, '"');
+        end = strchrW(end + 1, '"');
     }
     return NULL;
 }
@@ -157,7 +167,7 @@ static HANDLE run_process(void)
     }
     else
     {
-        cmdline = wcschr(cmdline, ' ');
+        cmdline = strchrW(cmdline, ' ');
     }
     if (!cmdline)
     {
@@ -196,9 +206,9 @@ static HANDLE run_process(void)
         else
         {
             start = cmdline;
-            end = wcschr(start, ' ');
+            end = strchrW(start, ' ');
             if (!end)
-                end = wcschr(start, '\0');
+                end = strchrW(start, '\0');
             remainder = end;
         }
 
